markdown/events/2021/bytesize-12-template-sync.md

```diff
@@ -38,3 +38,94 @@ The talk will be presented on Zoom and live-streamed on YouTube:
+
+[3:00](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=180) The `dev` branch is where we do all the development work, so things change there all the time. When we’re ready for a release, we merge `dev` into `master`.
+
+[3:14](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=194) Finally, we have this kind of special branch called `TEMPLATE`, and that’s what I’m going to be talking about.
+
+[3:24](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=204) So this is one of the figures from the [nf-core paper](https://rdcu.be/clDKw) , and it describes how automated sync works from nf-core `TEMPLATE`. One of the main things we offer within the nf-core organisation is this pipeline template. You can do `nf-core create`, where it asks you a couple of questions i.e. a name, a description, an author, and it creates a minimal but functional pipeline. It has a lot of the boilerplate code that we think is best practice for a Nextflow pipeline. So it sets up lots of config stuff and it creates all the code structure into different files, it sets up automated tests and Github actions, continuous integration etc. But that `TEMPLATE` never stands still, we’re adding to it all the time, and it grows with the community and experience. So if we fix something in one of the pipelines, which is general, we try and bring that back into the main `TEMPLATE`. But of course once you’ve done `nf-core create`, your repository is then separate and no longer tracks along with the nf-core subject, and that’s what we’re trying to bring in here. So the trick we’ve figured out is that when we make your pipeline, you straightaway make your commits where you haven’t done any changes - it’s just a template. You have that initial `commit` with the `TEMPLATE`, and we branch at that point to a new branch called `TEMPLATE`, which has the same shared `commit`. You can then just forget about that `TEMPLATE` branch and carry on doing whatever you’re doing in `master` or `dev`, but when there’s a new release of the nf-core `TEMPLATE`, we create a new `commit` along this special `TEMPLATE` branch. Now we’ve got one `commit` in this to commit, and the only changes in this branch are the things we updated in the main `TEMPLATE`. This is always the case for the `TEMPLATE` branch, it only ever has the vanilla template from the main nf-core `TEMPLATE`, and the commits are only ever just updates from there as we do new releases of tools. When you get a new `commit`, then we can do a `pull request` from that, `branch` it into the `dev` branch so that anything that changed in the boilerplate code is brought into `dev`. But it never goes the other way, dev is never brought into `TEMPLATE`.
```

Comment:
```suggestion
[3:24](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=204) So this is one of the figures from the [nf-core paper](https://rdcu.be/clDKw), and it describes how automated sync works from nf-core `TEMPLATE`. One of the main things we offer within the nf-core organisation is this pipeline template. You can do `nf-core create`, where it asks you a couple of questions i.e. a name, a description, an author, and it creates a minimal but functional pipeline. It has a lot of the boilerplate code that we think is best practice for a Nextflow pipeline. So it sets up lots of config stuff and it creates all the code structure into different files, it sets up automated tests and Github actions, continuous integration etc. But that `TEMPLATE` never stands still, we’re adding to it all the time, and it grows with the community and experience. So if we fix something in one of the pipelines, which is general, we try and bring that back into the main `TEMPLATE`. But of course once you’ve done `nf-core create`, your repository is then separate and no longer tracks along with the nf-core template, and that’s what we’re trying to bring in here. So the trick we’ve figured out is that when we make your pipeline, you straightaway make your commits where you haven’t done any changes - it’s just a template. You have that initial `commit` with the `TEMPLATE`, and we branch at that point to a new branch called `TEMPLATE`, which has the same shared `commit`. You can then just forget about that `TEMPLATE` branch and carry on doing whatever you’re doing in `master` or `dev`, but when there’s a new release of the nf-core `TEMPLATE`, we create a new `commit` along this special `TEMPLATE` branch. Now we’ve got one `commit` in this to commit, and the only changes in this branch are the things we updated in the main `TEMPLATE`. This is always the case for the `TEMPLATE` branch, it only ever has the vanilla template from the main nf-core `TEMPLATE`, and the commits are only ever just updates from there as we do new releases of tools. When you get a new `commit`, then we can do a `pull request` from that, `branch` it into the `dev` branch so that anything that changed in the boilerplate code is brought into `dev`. But it never goes the other way, dev is never brought into `TEMPLATE`.
```